src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs

```diff
@@ -196,7 +206,48 @@ private static Project CreateProject(ProjectId projectId, Solution solution)
         /// </summary>
         public DocumentId? GetDocumentId(SyntaxTree? syntaxTree, ProjectId? projectId)
         {
-            return State.GetDocumentState(syntaxTree, projectId)?.Id;
+            return GetDocumentState(syntaxTree, projectId)?.Id;
+        }
+
+        internal DocumentState? GetDocumentState(SyntaxTree? syntaxTree, ProjectId? projectId)
+        {
+            if (syntaxTree != null)
+            {
+                // is this tree known to be associated with a document?
+                var documentId = DocumentState.GetDocumentIdForTree(syntaxTree);
+                if (documentId != null && (projectId == null || documentId.ProjectId == projectId))
+                {
+                    // does this solution even have the document?
+                    var projectState = GetProjectState(documentId.ProjectId);
+                    if (projectState != null)
+                    {
+                        var document = projectState.DocumentStates.GetState(documentId);
+                        if (document != null)
+                        {
+                            // does this document really have the syntax tree?
+                            if (document.TryGetSyntaxTree(out var documentTree) && documentTree == syntaxTree)
+                            {
+                                return document;
+                            }
+                        }
+                        else
+                        {
+                            var generatedDocument = _compilationState.TryGetSourceGeneratedDocumentStateForAlreadyGeneratedId(documentId);
```

Comment:
todo: we can now check of the document id is for an SG doc or not.  We shouldn't bother with this (or we should update TryGetSourceGeneratedDocumentStateForAlreadyGeneratedId to check this) .